### PR TITLE
Change rules of indenting for spaces

### PR DESCRIPTION
**Pull request by iwek7**
Hello,

even though #28724 did not find much traction i was toying around with editor code to the point when I decided it's worthwhile to create draft PR of this improvement if not for merge then for future reference. This PR fixes multiple issues with spaces indentation and brings more consistency into this mechanic. I'd like to point out that those 'small issues' are really big annoyance in my everyday work with godot - that's why I decided to take time to fix them.

### **FEATURES AND FIXES**
When talking about spaces indenting in Godot editor there are five main cases to consider - clicking backspace without selection, tab and shift+tab without selection and tab and shift+tab with selection. Each of this cases has unique (and somewhat duplicated) code that provides indenting mechanism. As you will see I extracted some common methods but reality is that what we need is only two functions that handle indenting and unindenting - that is however, distant future (that I will be more than happy to work on if this subject gains some traction and support).

At this point it is worth noting that even though I carried out pretty careful testing efforts this code is pretty complicated and it has some regression potential.

**All below examples use space indentation with 4 spaces indent size, my work did not concern any of tabs issues. All screenshots of old behavior come from Godot 3.1 (I did not notice any changes to this on master)**

### Backspace at selection
In Godot there is feature that if you hit backspace (with no selection) and  there is only indentation before cursor the text gets unindented by configured indentation_size. This feature is standard in any IDE but Godot's implementation had two issues:

- If cursor column is less then indentation_size feature does not work and spaces are removed one by one
- Number of removed spaces was always equal to indentation_size - this means that if you hit accidental space you cant delete it easily with backspace because it will move you back to one-off position of previous indentation level. To visualize see below gif where I only click backspace (notice the indentation level is off by two and it remains so until all spaces get delete from the line).
backspace_old

I fixed it - when unindenting code first deletes spaces up until closest full indentation level and then removes whole levels. Feature works too if cursor column is smaller than indentation_size. Notice how at first two spaces are removed and then at next backspace click whole indentation level disappears.

backspace_new

### Unindenting with shift+tab
It is very similar as in 'Backspace at selection'. Code was always removing number of spaces equal to indentation_size. Example below - as you can see line  22 is slightly off (two spaces more than correct indentation level) and shift tabbing does not bring it to round indentation level - it just removes 4 spaces in this case.
shift_tab_no_selection_old

I fixed it! Just as before - first we remove enough spaces so that cursor ends up at full indentation level and then we remove whole level as before:

shift_tab_no_selection_new

### Indenting with tab

Similar issue - tabbing was always adding four spaces to the line. Usually when indenting you want to move stuff at cursor to new indentation level. Old mechanic:
tab_no_selection_old

My 'improvement' (not total fix since it still depends on cursor position in relation to the closest indent level - maybe we should look at position of first non-whitespace character?). Please notice how we at first 'even up' word after cursor to nearest selection and only at next tabs we move by whole indentation level:
tab_no_selection_new

### Indenting selection with tab and unindenting with shift+tab

Now it is where fun begins. Select something (can be multi-line) and hit that tab. I was actually unable to find similar feature in any of IDEs I'm using (QT-Creator, Intellij IDEA, Sublime Text) since there tab or untab when having selection would delete it but Godot attempts to indent all lines with selection. I kept this feature and improved it a bit. Here are the issues with current implementation:

- Selection changes weirdly when using this mechanic.
- Cursor moves weirdly (it is because correct cursor offset was only supported for tabs!)
- We always only insert indent_size number of spaces at the beginning of the line not caring for correct indentation level.
 
Preview of old mechanic:
tab_selection_multiline_old
tab_selection_single_line_old

Similar stuff was going on for unindenting selection, Not aligning to closest indentation level, moving selection and cursor. See (the cursor moves is me just holding shift+tab!):

shift_tab_selection_single_line
shift_tab_selection_multiline_old

I fixed cursor and selection changes issue (it keeps on same selection all the time - I thought it is ok) and also move text at first to closest full indentation level at first. My version with single line selection:

indenting_selection_single_line

And with multiline:

indenting_multiline_new

To be honest this part of the feature was most difficult to implement as it is not so obvious how it should work.

I think adding or improving this PR would be huge QOL improvement for Godot which would bring its indentation mechanics if not on par then close to other modern IDEs. 

Thanks for reading. Have a nice day.:)

### CODE REVIEW

File: scene/gui/text_edit.cpp [hunk 1]
```diff
@@ -1691,18 +1696,24 @@ void TextEdit::indent_right() {
 	for (int i = start_line; i <= end_line; i++) {
 		String line_text = get_line(i);
 		if (indent_using_spaces) {
-			line_text = space_indent + line_text;
+			// we don't really care where selection is - we just need to know indentation level at the beginning of the line
+			int left = _find_first_non_whitespace_column_of_line(line_text);
+			int spaces_to_add = _calculate_spaces_till_next_right_indent(left);
+			// since we will add this much spaces we want move whole selection and cursor by this much
+			selection_offset = spaces_to_add;
+			for (int i = 0; i < spaces_to_add; i++)
```
Comment (YeldhamDev) [inline scene/gui/text_edit.cpp]: This shadows the previous `i`, change it to a `j`.

File: scene/gui/text_edit.cpp [hunk 2]
```diff
@@ -2504,40 +2546,32 @@ void TextEdit::_gui_input(const Ref<InputEvent> &p_gui_input) {
 				} else {
 					if (k->get_shift()) {
 
-						//simple unindent
+						// simple unindent
 						int cc = cursor.column;
 
-						const int len = text[cursor.line].length();
+						const int line_length = text[cursor.line].length();
```
Comment (YeldhamDev) [inline scene/gui/text_edit.cpp]: Nothing is using this variable.

File: scene/gui/text_edit.cpp [hunk 3]
```diff
@@ -2504,40 +2546,32 @@ void TextEdit::_gui_input(const Ref<InputEvent> &p_gui_input) {
...
-									}
-									n++;
-								}
-
-								if (n > 0) {
-									_remove_text(cursor.line, cc - n, cursor.line, cc);
-									if (cursor.column > left - n) // inside text?
-										cursor_set_column(MAX(0, cursor.column - n));
+								// spaces unindentation
+								int  spaces_to_remove = _calculate_spaces_till_next_left_indent(cc);
```
Comment (YeldhamDev) [inline scene/gui/text_edit.cpp]: Unnecessary extra space after the `int`.

File: scene/gui/text_edit.cpp [hunk 4]
```diff
@@ -2546,9 +2580,14 @@ void TextEdit::_gui_input(const Ref<InputEvent> &p_gui_input) {
...
 						}
 					} else {
-						//simple indent
+						// simple indent
 						if (indent_using_spaces) {
-							_insert_text_at_cursor(space_indent);
+							// insert only as much spaces as needed till next indentation level
+							int spaces_to_add = _calculate_spaces_till_next_right_indent(cursor.column);
+							String indent_to_insert = String();
+							for(int i = 0; i < spaces_to_add; i++)
```
Comment (YeldhamDev) [inline scene/gui/text_edit.cpp]: A space is needed after the `for`.

**Comment by akien-mga** [comment]
See [Code style guidelines](http://docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html) for details on formatting.

File: scene/gui/text_edit.cpp (hunk 1 above)
Comment (iwek7) [reply scene/gui/text_edit.cpp]: fixed

File: scene/gui/text_edit.cpp (hunk 2 above)
Comment (iwek7) [reply scene/gui/text_edit.cpp]: fixed

File: scene/gui/text_edit.cpp (hunk 3 above)
Comment (iwek7) [reply scene/gui/text_edit.cpp]: fixed

File: scene/gui/text_edit.cpp (hunk 4 above)
Comment (iwek7) [reply scene/gui/text_edit.cpp]: fixed

**Comment by akien-mga** [comment]
We discussed this in PR meeting today, and the feature seems good. @Paulb23 should have a look as he's among the most familiar with TextEdit code.

I see that those changes are specific to space-based indentation. Did you check that the indentation behaviour is equivalent when using tab-based indentation? And what about a mix of tabs and spaces, e.g. `<tab><space><space><cursor>print("heya")` and pressing <kbd>Backspace</kbd>?

**Comment by iwek7** [comment]
I did not test this code with mix but my guess is that it will improve mixed case when space indentation is active. I will verify it and provide feedback.

As a side-note is there any place where tests for this code could be added? I spent vast majority of time on this testing and considering various edge cases - it would be good to automate this process.

**Comment by iwek7** [comment]
So it seems that in your example if tabs indentation is on then spaces will be removed one by one but if spaces indentation is on then code assumes every indentation character is space and will attempt to even up to closest indent as if it was <space><space><space><cursor>. This seems pretty extreme edge case (who would use tabs and have spaces indentation on?) but if you think it should be fixed I can do it.

**Comment by akien-mga** [comment]
It sounds like it should be fine as is. If we run into some corner cases while using the feature, we can discuss how to handle them further.

**Comment by akien-mga** [comment]
Thanks!
